salt/states/hanamod.py

```diff
@@ -0,0 +1,443 @@
+
+    name
+        The name of the secondary node
+    remote_host
+        Primary node hostname
+    remote_instance
+        Primary node instance
+    replication_mode
+        Replication mode
+    operation_mode
+        Operation mode
+    sid
+        System id of the installed hana platform
+    inst
+        Instance number of the installed hana platform
+    password
+        Password of the installed hana platform user
+    '''
+
+    ret = {'name': name,
+           'changes': {},
+           'result': False,
+           'comment': ''}
+
+    if not __salt__['hana.is_installed'](sid, inst, password):
+        ret['comment'] = 'HANA is not installed properly with the provided data'
+        return ret
+
+    current_state = __salt__['hana.get_sr_state'](sid, inst, password)
+    running = __salt__['hana.is_running'](sid, inst, password)
+    #  Improve that comparison
+    if running and current_state.value == 2:
+        ret['result'] = True
+        ret['comment'] = 'HANA node already set as secondary and running'
+        return ret
+
+    if __opts__['test']:
+        ret['result'] = None
+        ret['comment'] = '{} would be registered as a secondary node'.format(name)
+        ret['changes']['secondary'] = name
+        return ret
+
+    try:
+        #  Here starts the actual process
+        if running:
+            __salt__['hana.stop'](sid, inst, password)
+        __salt__['hana.sr_register_secondary'](
+            name, remote_host, remote_instance,
+            replication_mode, operation_mode, sid, inst, password)
+        __salt__['hana.start'](sid, inst, password)
+        new_state = __salt__['hana.get_sr_state'](sid, inst, password)
+        ret['changes']['secondary'] = name
+        ret['comment'] = 'HANA node set as {}'.format(new_state.name)
+        ret['result'] = True
+        return ret
+
+    except exceptions.CommandExecutionError as err:
+        ret['comment'] = six.text_type(err)
+        return ret
+
+
+def sr_clean(name, force, sid, inst, password):
+    '''
+    Clean the current node system replication mode
+    name:
+        Just for logging purposes
+    force
+        Force cleanup process
+    sid
+        System id of the installed hana platform
+    inst
+        Instance number of the installed hana platform
+    password
+        Password of the installed hana platform user
+    '''
+
+    ret = {'name': name,
+           'changes': {},
+           'result': False,
+           'comment': ''}
+
+    if not __salt__['hana.is_installed'](sid, inst, password):
+        ret['comment'] = 'HANA is not installed properly with the provided data'
+        return ret
+
+    current_state = __salt__['hana.get_sr_state'](sid, inst, password)
+    running = __salt__['hana.is_running'](sid, inst, password)
+    #  Improve that comparison
+    if current_state.value == 0:
+        ret['result'] = True
+        ret['comment'] = 'HANA node already clean'
+        return ret
+
+    if __opts__['test']:
```

Comment:
I understand your point, anyway the official contributing docs suggest to do how I did (that's why I implemented this way): https://docs.saltstack.com/en/latest/ref/states/writing.html

If you still think that we should change I don't have any problem with that.

Reply:
@arbulu89 yeah, I know about that doc. And always wondering if this test mode the way it is implemented was at all good idea at all. :wink: I am more in favour how Terraform is doing, but that's probably way too much of discussion in this context here.

My _personal_ reasoning here is that Hana, known as heavyweight software, probably could be better not touched, if not needed. And testing is only describing what is going to be done. Like dry-run, where you do not really running anything, but forecasting. So lean to only say "1, 2 and 3 will be applied". But the _result_ if this is applied or not — something else. Say, you wanted Hana running. So you say "Hana will be running". Then you go, actually check it, see it is already running. So you then only saying "Hana is already running, no changes has been made". And so your test dry-run is valid: Hana is running (at the end). This was my poorly expressed example, hopefully you understand what I mean.

But it is indeed interesting what to do here. @meaksh @brejoc @dincamihai thoughts?

Reply:
@isbm if I understand correctly, in this case, when "test" is passed in the options, no change would be applied.
Indeed, it would check if hana.is_installed, but are we expecting that this would apply any changes?
In case hana.is_installed would raise an exception, wouldn't the displayed message be relevant?